File: baml_language/sdk_tests/crates/swift/function_calls/customizable/TestStdlibEntrypoints.swift
// Stdlib entrypoints — port of python_pydantic2
// `test_stdlib_entrypoints.py`: native ($rust_function) and SysOp
// ($rust_io_function) stdlib functions callable as ordinary generated
// entry points.
import XCTest
import Foundation
import Baml

final class TestStdlibEntrypoints: XCTestCase {
    // `baml.sys.argv() -> string[]` is a `$rust_function` ->
    // `FunctionKind::Native`. The fixture host passes no program arguments, so
    // the contents are not worth asserting on — that the call lands and is
    // stable across invocations is.
    func test_stdlib_entrypoints_native_argv_callable_as_entry_point() throws {
        let args = try Baml.baml.sys.argv()
        XCTAssertEqual(args, try Baml.baml.sys.argv())
    }

    func test_stdlib_entrypoints_sysop_fs_exists_callable_as_entry_point() throws {
        XCTAssertTrue(try Baml.baml.fs.exists(path: "."))
    }

    func test_stdlib_entrypoints_compiler_intrinsics_are_not_emitted_as_entry_points() throws {
        // Python inspects generated .py files; the Swift analog reads
        // the generated sources. `vendor.log.*` and `baml.events.send`
        // are compiler intrinsics and must not surface as entry points.
        let sourcesDir = URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent()  // Tests/BamlTests
            .deletingLastPathComponent()  // Tests
            .deletingLastPathComponent()  // generated
            .appendingPathComponent("Sources/Baml")

        let vendorSource = try? String(
            contentsOf: sourcesDir.appendingPathComponent("vendor.swift"),
            encoding: .utf8
        )
        if let vendorSource {
            for forbidden in ["func info(", "func debug(", "func warn(", "func error("] {
                XCTAssertFalse(
                    vendorSource.contains(forbidden),
                    "compiler intrinsic leaked into vendor.swift: \(forbidden)"
                )
            }
        }

        let bamlSource = try String(
            contentsOf: sourcesDir.appendingPathComponent("baml.swift"),
            encoding: .utf8
        )
        XCTAssertFalse(
            bamlSource.contains("\"baml.events.send\""),
            "compiler intrinsic baml.events.send leaked into baml.swift"
        )
    }
}

File: baml_language/sdk_tests/crates/java/function_calls/customizable/TestStdlibEntrypoints.java
// Stdlib functions of different `FunctionKind`s are callable as entry points
// (directly from the host), not only from inside BAML.
//
// Port of python_pydantic2/function_calls/customizable/test_stdlib_entrypoints.py.
//
// java-port note: BAML `baml.sys` / `baml.fs` namespaces map to packages
// `baml_sdk.baml.sys` / `baml_sdk.baml.fs`, each with its own `Fns` holder.
// Two `Fns` classes can't both be imported (no import aliases in Java), so
// they are fully qualified at the call site.

import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import org.junit.jupiter.api.Test;

class TestStdlibEntrypoints {

    // Intrinsic-only modules are not emitted at all, so a missing file is fine;
    // callers only need to confirm the symbol is absent when the file exists.
    private static String generatedSdkFile(String relPath) throws IOException {
        Path path = Paths.get("baml_sdk", relPath);
        if (!Files.exists(path)) {
            return null;
        }
        return Files.readString(path);
    }

    // `baml.sys.argv() -> string[]` is a `$rust_function` ->
    // `FunctionKind::Native`. Calling it as an entry point should run the
    // native and return the argument list, not reject with
    // `NotInvokableAsEntry`. The fixture host passes no program arguments, so
    // the contents are not worth asserting on — that the call lands and
    // returns a list is.
    @Test
    void test_stdlib_entrypoints_native_argv_callable_as_entry_point() {
        assertNotNull(baml_sdk.baml.sys.Fns.argv());
    }

    // `baml.fs.exists(path: string) -> bool` is a `$rust_io_function` ->
    // `FunctionKind::SysOp`. Calling it as an entry point should run the
    // filesystem sysop and return a bool. `.` exists in the generated fixture
    // directory on the test host.
    @Test
    void test_stdlib_entrypoints_sysop_fs_exists_callable_as_entry_point() {
        assertTrue(baml_sdk.baml.fs.Fns.exists("."));
    }

    @Test
    void test_stdlib_entrypoints_compiler_intrinsics_are_not_emitted_as_entry_points() throws IOException {
        // java-port note: Python inspects both the `.py` (wire FQN string
        // literals) and the `.pyi` (stub `def` decls). Java emits a single
        // `Fns.java` per namespace, so each row checks that holder for either
        // the wire FQN literal or the generated method declaration.
        String[][] forbidden = {
            {"vendor/log/Fns.java", "\"log.info\""},
            {"vendor/log/Fns.java", "\"log.debug\""},
            {"vendor/log/Fns.java", "\"log.warn\""},
            {"vendor/log/Fns.java", "\"log.error\""},
            {"vendor/log/Fns.java", "info("},
            {"vendor/log/Fns.java", "debug("},
            {"vendor/log/Fns.java", "warn("},
            {"vendor/log/Fns.java", "error("},
            {"baml/events/Fns.java", "\"baml.events.send\""},
            {"baml/events/Fns.java", "send("},
        };

        for (String[] entry : forbidden) {
            String contents = generatedSdkFile(entry[0]);
            assertTrue(contents == null || !contents.contains(entry[1]));
        }
    }
}
